### ISSUE TITLE
Can keras create a "deep dream"?

### ISSUE THREAD

**Issue by paipai880429** [fchollet/keras#318]
Google Research has just published there up-to-date achievements in deep learning http://googleresearch.blogspot.ch/2015/06/inceptionism-going-deeper-into-neural.html. They have created beautiful arts by leveraging ConvNets. Yesterday, they also make their code public available https://github.com/google/deepdream which is based on **Caffe**. I am just wondering if we can create such arts based on **keras**?

**Comment by fchollet** [comment]
Now that the technique is open source, yes, absolutely, you can reimplement it with Keras.

> Making the "dream" images is very simple. Essentially it is just a gradient ascent process that tries to maximize the L2 norm of activations of a particular DNN layer. 

In fact, you can probably reuse most of the Python code they provide. https://github.com/google/deepdream/blob/master/dream.ipynb

This is left as an exercise to the reader.

**Comment by mthrok** [comment]
Looking at their deploy model (https://github.com/BVLC/caffe/blob/master/models/bvlc_googlenet/deploy.prototxt), what are used there are `convolution` (with stride),`max pooling`, `relu`, `concat`(`merge` in keras I think?), `LRN`, `dropout`, `softmax`, `inner_product` (`dense` in keras). 

I think recently `LRN` PR was created (or already merged?) and there is ongoing discussion on converting caffe model to keras data. Thankfully pycaffe offers net surgery tool which should enable us to convert caffe model to keras. So I think its possible.

**Comment by fchollet** [comment]
As a final follow-up: we now have a Deep Dream script in the examples folder: https://github.com/fchollet/keras/blob/master/examples/deep_dream.py

**Comment by paipai880429** [comment]
Thanks for providing us such a beautiful tool!

**Comment by paipai880429** [comment]
t

**Comment by meanmee** [comment]
How to acces the trained VGG_16.weights for deep dream?

**Comment by baraldilorenzo** [comment]
@meanmee here: https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3

**Comment by danielvarga** [comment]
@baraldilorenzo @meanmee @pengpaiSH @fchollet At https://github.com/danielvarga/keras-deep-dream I've created an implementation based on the example code, with two extra features:

- it can work with any of the cool new https://github.com/fchollet/deep-learning-models
- octaves (iterative scaling) are implemented.

Work in progress.
